fix: Fix file/email attachment name containing percent sign

cogk opened the pull request:
Needs careful review :mag: 

## Summary

The path to a saved file on disk, also known as `file_url`, must not contain `%` to be accessible through Nginx. Indeed, a browser trying to access a file named `%.png` will ask for `/files/%25.png` which will look for the file `/files/%.png` as expected. However, this file won't exist if the `file_url` is not unquoted beforehand (i.e. invalidly saved as `file_url` = `/files/%25.png`). This did happen in some cases.

## Assumptions

When fixing the bugs mentioned below, I assumed the following:
- When creating a file using the `file_name` + `content` parameters, I expect the file name on disk to be the same: `%50 off promo.pdf` stays the same. I will explain why the percent sign specifically cannot remain unchanged in a file name.
- When creating a file using the `file_url`, we want to unquote beforehand because we assume that it is percent-encoded: for example, I copy-paste a link from my browser into the File upload box for external links in Frappe.

## Bugs fixed

This was the case because of a combination of bugs:
- extracting files from inbound emails did not unquote percent-encoded file names (provided in the Content-Type header).
- the `file_url` was correctly unquoted, but every time the file was saved (on `validate`), leading to repeated unquotings.
- the `file_name` could contain intended percent signs. A file name like `50% off promo code.pdf`), did lead to a `file_url` containing a percent sign, which cannot be accessed when the encoding is ambiguous (`%50`)

---

Feel free to challenge the assumptions, fixes, and ask questions.

cogk committed:
fix(File): Unquote file_url only once

cogk committed:
fix(File): Disallow percent sign in file_name

cogk committed:
fix(InboundEmail): Unquote percent-encoded file names

cogk committed:
test(File): Add test for attachments with percent-encoded file names

cogk commented:
- [x] check if semgrep is right

barredterra commented:
@mergifyio update

barredterra commented:
Looks like semgrep is right, but it's not your fault. I think the reported problems should be addressed in a separate PR.

cogk commented:
up

akhilnarang commented:
@mergifyio rebase

cogk commented:
my lovely stalebot, there was activity 😢

cogk commented:
thanks @akhilnarang

akhilnarang commented:
@cogk could you have a look at the failing test? I'm fine to merge it then.

cogk committed:
fix(file): Remove illegal chars in file_name

- / and \ cannot work as file names
- % is not allowed verbatim in URLs
- ? and # are not allowed because confused with query/hash separators

cogk commented:
All checks are green, PR is ready.

akhilnarang commented:
@mergify backport version-15-hotfix
